File: python/149.py
"""Project Euler Problem 149 Solution.

Generate a 2000x2000 table using the specified Lagged Fibonacci Generator
and find the maximum sum of adjacent entries in any direction (horizontal,
vertical, diagonal, or anti-diagonal).
"""

from typing import List

# Constants for clarity and maintainability
MOD = 1_000_000
OFFSET = 500_000
N = 2000
TOTAL_CELLS = N * N


def kadane(arr: List[int]) -> int:
    """Kadane's algorithm for 1D maximum subarray sum.
    
    Handles empty arrays and all-negative cases correctly.
    """
    if not arr:
        return -float('inf')

    max_ending_here = max_so_far = arr[0]
    for i in range(1, len(arr)):
        max_ending_here = max(arr[i], max_ending_here + arr[i])
        max_so_far = max(max_so_far, max_ending_here)
    return max_so_far


def generate_sequence() -> List[int]:
    """Generate the Lagged Fibonacci sequence."""
    s: List[int] = [0] * TOTAL_CELLS

    # Generate first 55 values (1-based indices 1 to 55, 0-based indices 0 to 54)
    for i in range(55):
        k = i + 1
        temp = (100003 - 200003 * k + 300007 * k**3) % MOD
        s[i] = temp - OFFSET

    # Generate remaining values (1-based indices 56 to 4000000, 0-based 55 to 3999999)
    for i in range(55, TOTAL_CELLS):
        k1 = s[i - 24]
        k2 = s[i - 55]
        # Add MOD to ensure non-negative before modulo operation
        temp = (k1 + k2 + MOD) % MOD
        s[i] = temp - OFFSET

    return s


def main() -> int:
    """Main function."""
    s = generate_sequence()

    # Create the 2D table from the 1D sequence (row-major order)
    table: List[List[int]] = []
    for i in range(N):
        row = s[i * N:(i + 1) * N]
        table.append(row)

    max_sum = -float('inf')

    # Check horizontal lines
    for row in table:
        max_sum = max(max_sum, kadane(row))

    # Check vertical lines
    for col in range(N):
        column = [table[row][col] for row in range(N)]
        max_sum = max(max_sum, kadane(column))

    # Check main diagonals (top-left to bottom-right)
    for start_row in range(N):
        diag = []
        r, c = start_row, 0
        while r < N and c < N:
            diag.append(table[r][c])
            r += 1
            c += 1
        max_sum = max(max_sum, kadane(diag))

    for start_col in range(1, N):
        diag = []
        r, c = 0, start_col
        while r < N and c < N:
            diag.append(table[r][c])
            r += 1
            c += 1
        max_sum = max(max_sum, kadane(diag))

    # Check anti-diagonals (top-right to bottom-left)
    for start_col in range(N):
        diag = []
        r, c = 0, start_col
        while r < N and c >= 0:
            diag.append(table[r][c])
            r += 1
            c -= 1
        max_sum = max(max_sum, kadane(diag))

    for start_row in range(1, N):
        diag = []
        r, c = start_row, N - 1
        while r < N and c >= 0:
            diag.append(table[r][c])
            r += 1
            c -= 1
        max_sum = max(max_sum, kadane(diag))

    return max_sum


if __name__ == "__main__":
    print(main())

File: c/149.c
/*
 * Project Euler 149 - Maximum-sum subsequence in a 2000x2000 table
 *
 * Generate table with Lagged Fibonacci Generator, then find max sum
 * contiguous subsequence in any row, column, or diagonal (Kadane's algorithm).
 */
#include <stdio.h>
#include <stdlib.h>

#define N 2000
#define TOTAL (N * N)
#define MOD 1000000
#define OFFSET 500000

static int table[N][N];

static int kadane(int *arr, int len) {
    if (len == 0) return -2147483647;
    int max_here = arr[0], max_so_far = arr[0];
    for (int i = 1; i < len; i++) {
        if (max_here + arr[i] > arr[i])
            max_here = max_here + arr[i];
        else
            max_here = arr[i];
        if (max_here > max_so_far)
            max_so_far = max_here;
    }
    return max_so_far;
}

int main(void) {
    /* Generate sequence */
    int *s = malloc(TOTAL * sizeof(int));

    for (int i = 0; i < 55; i++) {
        long long k = i + 1;
        long long temp = (100003 - 200003*k + 300007*k*k*k) % MOD;
        if (temp < 0) temp += MOD;
        s[i] = (int)(temp - OFFSET);
    }

    for (int i = 55; i < TOTAL; i++) {
        int temp = (s[i-24] + s[i-55] + MOD) % MOD;
        s[i] = temp - OFFSET;
    }

    /* Fill table */
    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            table[i][j] = s[i * N + j];

    free(s);

    int max_sum = -2147483647;
    int *buf = malloc(N * sizeof(int));

    /* Horizontal */
    for (int i = 0; i < N; i++) {
        int v = kadane(table[i], N);
        if (v > max_sum) max_sum = v;
    }

    /* Vertical */
    for (int j = 0; j < N; j++) {
        for (int i = 0; i < N; i++) buf[i] = table[i][j];
        int v = kadane(buf, N);
        if (v > max_sum) max_sum = v;
    }

    /* Main diagonals (top-left to bottom-right) */
    /* Starting from row r, col 0 */
    for (int r = 0; r < N; r++) {
        int len = 0;
        int i = r, j = 0;
        while (i < N && j < N) {
            buf[len++] = table[i][j];
            i++; j++;
        }
        int v = kadane(buf, len);
        if (v > max_sum) max_sum = v;
    }
    /* Starting from row 0, col c */
    for (int c = 1; c < N; c++) {
        int len = 0;
        int i = 0, j = c;
        while (i < N && j < N) {
            buf[len++] = table[i][j];
            i++; j++;
        }
        int v = kadane(buf, len);
        if (v > max_sum) max_sum = v;
    }

    /* Anti-diagonals (top-right to bottom-left) */
    /* Starting from row 0, col c */
    for (int c = 0; c < N; c++) {
        int len = 0;
        int i = 0, j = c;
        while (i < N && j >= 0) {
            buf[len++] = table[i][j];
            i++; j--;
        }
        int v = kadane(buf, len);
        if (v > max_sum) max_sum = v;
    }
    /* Starting from row r, col N-1 */
    for (int r = 1; r < N; r++) {
        int len = 0;
        int i = r, j = N - 1;
        while (i < N && j >= 0) {
            buf[len++] = table[i][j];
            i++; j--;
        }
        int v = kadane(buf, len);
        if (v > max_sum) max_sum = v;
    }

    free(buf);
    printf("%d\n", max_sum);
    return 0;
}
